ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java

```diff
@@ -183,24 +183,31 @@ public Void visitLikePredicate(final LikePredicate node, final TypeContext conte
 
     @Override
     public Void visitFunctionCall(final FunctionCall node, final TypeContext context) {
-      final List<SqlArgument> argumentTypes = new ArrayList<>();
       final FunctionName functionName = node.getName();
+
+      final TypeContext contextCopy = context.getCopy();
+      final List<SqlArgument> argumentTypes = new ArrayList<>();
       final boolean hasLambda = node.hasLambdaFunctionCallArguments();
       for (final Expression argExpr : node.getArguments()) {
-        final TypeContext childContext = context.getCopy();
+        final TypeContext childContext;
+        if (argExpr instanceof LambdaFunctionCall) {
+          childContext = contextCopy.getCopy();
+        } else {
+          childContext = context.getCopy();
+        }
         final SqlType resolvedArgType =
-            expressionTypeManager.getExpressionSqlType(argExpr, childContext);
-        process(argExpr, context.getCopy());
+            expressionTypeManager.getExpressionSqlType(argExpr, childContext.getCopy());
```

Comment:
We are a bit duplicate copying context here. Maybe we can wrap the logic of

```
        final TypeContext childContext;
        if (argExpr instanceof LambdaFunctionCall) {
          childContext = contextCopy.getCopy();
        } else {
          childContext = context.getCopy();
        }
```

as a `argTypeContext(argExpr, parentContext, lambdaSharedContext)`, which would be called for both `expressionTypeManager#getExpressionSqlType` and `process`? Ditto on other classes.